soc/arm/silabs_sim3/sim3u/gen_crossbar_config.py

```diff
@@ -0,0 +1,343 @@
+#!/usr/bin/env python3
+#
+# Copyright (c) 2023 GARDENA GmbH
+#
+# SPDX-License-Identifier: Apache-2.0
+
+"""Generate crossbar register values from devicetree pinctrl nodes.
+
+The sim3u soc does support pinmuxing, but with many limimitations.
+It's also non-optional so we can't just not support it and rely on default
+values.
+
+The hardware doesn't allow us to properly implement the pinmux API, but
+we still want to use the standard pinctrl nodes inside the devicetree to define
+which peripheral should be muxed to which pins. To accomplish that, this script
+parses these nodes and generates code that applies all `default` nodes of all
+enabled devices within soc.c.
+
+There's two crossbars:
+- crossbar0: which controls portbanks 0 and 1
+- crossbar1: which controls portbanks 2 and 3
+
+Each crossbar has two configuration-values which reside in different registers:
+- config: A bitmask which tells the crossbar which peripherals should be muxed.
+          Some peripherals have multiple bits to prevent having to mux unused
+          pins (like UART flow control).
+- enable: A bit which enables or disables the whole crossbar. When disabled,
+          all pins of the crossbar are disconnected.
+And each portbank has this related config:
+- pbskipen: A bitmask where value `1` means, that the pin will not be muxed.
+            The index of the bit refers to the pin number.
+
+A crossbar has a list of signals that it tries to mux to pins that it controls.
+That list has a fixed order and signals that belong to the same peripheral are
+next to each other. The crossbar hardware simply iterates the signal list and
+assigns each signal to the lowest available pin. That has a few implications:
+- Pins of a peripheral are always consecutive within all non-skipped pins.
+- Peripherals that appear first in the signal list, will always use lower
+  pin-numbers than later ones
+- There's no way to change the muxing without side-effects. Writing to pbskipen
+  while the crossbar is enabled might temporarily cause unwanted muxing.
+  Disabling the crossbar will disconnect all peripherals for a short time.
+
+The last point is the reason why we don't implement zephyrs runtime pinmuxing
+API. Applying the pinmuxing one by one as drivers get initialized would require
+disconnecting pins quite often.
+"""
+
+import sys
+import pickle
+import argparse
+import os
+import enum
+
+sys.path.insert(
+    0, os.path.join(os.environ["ZEPHYR_BASE"], "scripts/dts/python-devicetree/src")
+)
+
+from devicetree import edtlib
+
+
+class Signal(enum.Enum):
+    USART0_TX = 0
+    USART0_RX = 1
+    USART0_RTS = 2
+    USART0_CTS = 3
+    USART0_UCLK = 4
+
+    USART1_TX = 9
```

Comment:
can you please explain why we need pinctrl related Python scripts in `soc/`? If this is required by the HAL, -1 to have this in tree.

Reply:
This is described in detail in the documentation of said python module: https://github.com/zephyrproject-rtos/zephyr/pull/64209/files#diff-0a14224d1df8e0d15ef02b23405ccca2fa3ee71eb65671d7c03b289357a112beR7

This is a workaround due to hardware limitations. The pinctrl hardware functions under the assumption that you set all the configs once at boot and never touch them again. This doesn't fit zephyrs pinctrl subsystem which tries to configure pins for every peripheral on demand, one by one. The python script allows using devictree pinctrl nodes to simplify pin configuration even though we do not support the pinctrl subsystem.

Reply:
can you please provide a datasheet or something to look at?

Reply:
https://www.silabs.com/documents/public/data-sheets/SiM3U1xx-SiM3C1xx-RM.pdf
`8. Port I/O Configuration`, Specifically the crossbar descriptions.

Reply:
I guess it's

> When configuring the crossbar, all settings should be made to the crossbar and Port Bank registers before
enabling the crossbar. This ensures that peripherals will not shift around while each one is being enabled and Port
I/O pins will remain stable. The settings in PBOUTMD, PBMDSEL, or PBSKIPEN will not take effect until the
crossbars are enabled.

right?

Reply:
exactly.